bip-taro.mediawiki

```diff
@@ -219,21 +225,134 @@ Taro-specific commitments, then the same or different leaf version can be used
+
+Due to the lexicographical sorting of tap branches/leaves before hashing, we
+lose the ordering information within the merkle tree. As a result, we're
+instead forced to require reveal of the sibling pre-image.
+
+The following algorithm can be used to asset the uniqueness of the Taro
+commitment in a tapscript tree:
+<source lang="python">
+taro_commitment_is_valid(commitment_output: TxOut, control_block: ControlBlock, 
+    sibling_preimage: TaroTapReveal, taro_root: []byte) -> bool
+    
+    taro_root_hash = tagged_hash("TapLeaf", taro_root)
+    internal_key = parse_key(control_block.internal_key_bytes)
+
+    match len(control_block.size):
+       # Just the internal public key, so we can just verify the commitment below.
+       case 32:
+           expected_output_key = internal_key + tagged_hash("TapTweak", internal_key || taro_root_hash)
+
+           return expected_output_key == witness_program(commitment_output.pk_script)
+
+       # More than one element in the tree, so need to verify each sub-case.
+       case 65:
+           
+           match sibling_preimage:
+               # The pre-image is a leaf, so verify the hashes match up.
+               case LeafReveal(leaf_bytes):
+                   leaf_hash = tagged_hash("TapLeaf", leaf_bytes)
+
+                   if leaf_hash != control_block.sibling_hash:
+                       return false
+
+               # The pre-image is itself a branch, so we need the two siblings
+               # used to construct the branch.
+               case BranchReveal(sibling_1, sibling_2):
+                   expected_branch_hash = tagged_hash("TapBranch", sort(sibling_1, sibling_2))
+
+                   if expected_branch_hash != construct.sibling_hash:
+                       return false
+
+               # We know the tree is well formed now, so we'll verify the root
+               # commitment.
+               root_hash = tagged_hash("TapBranch", sort(taro_root_hash, control_block.sibling_hash))
+
+               expected_output_key = internal_key + tagged_hash("TapTweak", internal_key || root_hash)
+
+               return expected_output_key == witness_program(commitment_output.pk_script)
+
+       default:
+           return false
+
+</source>
+
+In addition to verifying that a commitment is unique within the constraints of
+a single taproot tree, we also need to verify that there is no similar
+commitment across the remaining outputs in the genesis transaction. We can do
+this by asserting that:
+* For each output on the genesis transaction that isn't creating an asset:
+** If top level key was derived using BIP 86, then verify the key derivation.
+** If the key commits to a script path:
+*** If only a single element is in the tree, verify that isn't a duplicate commitment.
+*** If the tree contains more than one element, then verify that the pre-image of the two branches isn't a taro commitment.
```

Comment:
Are we checking only for duplicate commitments, or any other commitments? Here (and in the pseudocode) we allow non-duplicate commitments if they are the only tapscript element, but disallow them at depth 1.

Reply:
So it's sort of implied by the `commitment_expected` call that we're mainly ensuring that none of the "expected" outputs have a commitment here. As far as allowing non-duplicate commitments (other valid taro commitments), that's accepted here as long as it isn't committing to that specific taro root. 

I think I don't fully understand your questions however?

Reply:
Or actually, I think you're getting at the fact that assuming `commitment_expected` knows which outputs to expect a valid commitment, of we also need to check for the marker byte prefix here of this leaf hash?

Reply:
I was confused about if we are allowing multiple taro commitments in the same TX in general. Regardless of if we are or aren't, the rules / checks should be the same at depth 0 and depth 1.

Here the checks are written as:
-At depth 0:
--verify that isn't a duplicate commitment
-At depth 1:
--verify that the pre-image of the two branches isn't a taro commitment

If we want to allow multiple valid taro commitments, then we should relax the checks at depth 1 / in the MultiLeaf case to only assert that both branch preimages are not duplicate taro commitments.

Reply:
> I was confused about if we are allowing multiple taro commitments in the same TX in general

A single _output_ should only have a single commitment. Other outputs may have commitments as well depending on the on chain event (like a transfer for example will produce a transaction with multiple commitments).

Reply:
For asset creation, the asset ID includes the output index, so this check here is just wanting to ensure that the verifier is aware of any relevant assets being issued in the transaction. 

For transfers, this is needed to ensure that they aren't committing to a history where the transfer didn't happen.

Reply:
Cleared this up offline, will be more explicit here w.r.t what we're trying to verify and also the intent of `commitment_expected`.